src/NovelStepTimer.h

```diff
@@ -0,0 +1,125 @@
+//
+// Created by tagoo on 12/09/2019.
+//
+
+// This is based on the StepTimer provided in the DirectX ToolKit
+// Original code is available under the MIT License
+
+#ifndef NOVELRT_STEPTIMER_H
+#define NOVELRT_STEPTIMER_H
+
+#include <stdint.h>
+#include <vector>
+
+namespace NovelRT {
+  typedef void (*NovelUpdateSubscriber)(const double deltaSeconds);
+
+  class StepTimer {
+  public:
+    // Windows and some other platforms use 100ns ticks
+    static const uint64_t TicksPerSecond = 10'000'000;
+
+    StepTimer(uint32_t targetFrameRate = 0);
+
+    uint64_t getElapsedTicks() const;
+    uint64_t getTotalTicks() const;
+
+    double getElapsedSeconds() const;
+    double getTotalSeconds() const;
+
+    uint64_t getTargetElapsedTicks() const;
+    void setTargetElapsedTicks(uint64_t value);
+
+    double getTargetElapsedSeconds() const;
+    void setTargetElapsedSeconds(double value);
+
+    uint32_t getFrameCount() const;
+    uint32_t getFramesPerSecond() const;
+
+    bool getIsFixedTimeStep() const;
+    void setIsFixedTimeStep(bool value);
+
+    static double TicksToSeconds(uint64_t ticks) {
+      return (double)ticks / TicksPerSecond;
+    }
+
+    static uint64_t SecondsToTicks(double seconds) {
+      return (uint64_t)(seconds * TicksPerSecond);
+    }
+
+    void resetElapsedTime();
+    void tick(std::vector<NovelUpdateSubscriber> update);
+
+  private:
+    const uint64_t _frequency;
+    const uint64_t _maxCounterDelta;
+
+    uint64_t _lastCounter;
+    uint64_t _secondCounter;
+    uint64_t _remainingTicks;
+
+    uint64_t _elapsedTicks;
```

Comment:
Most values are tracked in terms of "ticks" which are unsigned integers. This ensures rounding error doesn't need to be thought about and results are always as accurate as possible.